src/piecewise.jl

```diff
@@ -0,0 +1,208 @@
+struct Domain{L, R, T, S}
+    lo::T
+    hi::S
+end
+
+Domain{L, R}(lo::T, hi::S) where {T, S, L, R} = Domain{L, R, T, S}(lo, hi)
+Domain(lo, hi) = Domain{:open, :closed}(lo, hi)
+Domain(lo::Tuple, hi::Tuple) = Domain{lo[2], hi[2]}(lo[1], hi[1])
+Domain(X::Interval) = Domain{:closed, :closed}(inf(X), sup(X))
+Domain() = Domain{:open, :open}(Inf, -Inf)
+
+lowerbound(x::Domain{L, R}) where {L, R} = (x.lo, L)
+upperbound(x::Domain{L, R}) where {L, R} = (x.hi, R)
+
+inf(x::Domain) = x.lo
+sup(x::Domain) = x.hi
+
+"""
+    rightof(val, lowerbound)
+
+Determine if a value is on the right of a lower bound.
+"""
+function rightof(x::Real, (val, bound))
+    bound == :closed && return val <= x
+    return val < x
+end
+
+function rightof((val1, bound1), (val2, bound2))
+    if val1 < val2 || (val1 == val2 && bound1 == bound2 == :closed)
+        return false
+    end
+
+    return true
+end
+
+"""
+    leftof(val, upperbound)
+
+Determine if a value is on the left of an upper bound.
+"""
+function leftof(x::Real, (val, bound))
+    bound == :closed && return x <= val
+    return x < val
+end
+
+function leftof((val1, bound1), (val2, bound2))
+    if val2 < val1 || (val1 == val2 && bound1 == bound2 == :closed)
+        return false
+    end
+
+    return true
+end
+
+function leftof(d1::Domain, d2::Domain)
+    val1, bound1 = upperbound(d1)
+    val2, bound2 = lowerbound(d2)
+
+    val1 == val2 && return !(bound1 == bound2 == :closed)
+    return val1 < val2
+end
+
+Base.in(x::Real, domain::Domain) = rightof(x, lowerbound(domain)) && leftof(x, upperbound(domain))
```

Comment:
Mhm maybe we should only export `Domain` to the user and not overload Base functions for it.

Reply:
Why not ? They seem quite natural operation to perform.

Reply:
I am not entirely sure 😄. My reasoning is based on maintaining consistency : it feels somewhat inconsistent that we have `in_interval` for `BareInterval`, while for `Domain`, we overload `Base.in`. This is maybe a "flaw" or something to be improved, I do not know....
However, for now, it seems more prudent to first implement the necessary internal functionalities and then extend them incrementally on 0.22.x (this way we also avoid having to deal with breaking changes).

Reply:
A `Domain` is a set, and never a number, so we don't have the same ambiguity we have for an `Interval`. I can't see how it could go wrong (famous words to regrets, I know ^^).